rust/src/storage/file/mod.rs

```diff
@@ -87,33 +84,70 @@ impl StorageBackend for FileStorageBackend {
     async fn list_objs<'a>(
         &'a self,
         path: &'a str,
-    ) -> Result<
-        Pin<Box<dyn Stream<Item = Result<ObjectMeta, StorageError>> + Send + 'a>>,
-        StorageError,
-    > {
-        let readdir = ReadDirStream::new(fs::read_dir(path).await?);
-
-        Ok(Box::pin(readdir.err_into().and_then(|entry| async move {
-            let path = String::from(
-                entry
-                    .path()
-                    .to_str()
-                    .ok_or_else(|| StorageError::Generic("invalid path".to_string()))?,
-            );
-            match entry.metadata().await {
-                Ok(meta) => {
-                    Ok(ObjectMeta {
-                        path,
-                        modified: meta.modified()?.into(),
-                        size: Some(meta.len().try_into().map_err(|_| {
-                            StorageError::Generic("cannot convert to i64".to_string())
-                        })?),
-                    })
+    ) -> Result<BoxStream<'a, Result<ObjectMeta, StorageError>>, StorageError> {
+        let walkdir = WalkDir::new(path)
+            // Don't include the root directory itself
+            .min_depth(1);
+
+        let s =
+            walkdir.into_iter().flat_map(move |result_dir_entry| {
+                match convert_walkdir_result(result_dir_entry) {
+                    Err(e) => Some(Err(e)),
+                    Ok(None) => None,
+                    Ok(entry @ Some(_)) => entry
+                        .filter(|dir_entry| dir_entry.file_type().is_file())
+                        .map(|entry| {
+                            let file_path =
+                                String::from(entry.path().to_str().ok_or_else(|| {
```

Comment:
This might be bike shedding a bit, but might be nice to have this as a match guard and save a level of indentation. (Note the suggestion requires additional edits to compile.)

```suggestion
                    Ok(Some(entry)) if entry.file_type().is_file() => {
                        let file_path =
                            String::from(entry.path().to_str().ok_or_else(|| {
```

Reply:
I tried to make this flatter, but as is turns out, when doing so, we have to wrap the result returned in this branch in an option. This in turn lead to having to internally deal with all the fallible operations, since we no longer can implicitly convert as we are returning an `Option<Result<..>>`. Tried several permutations, but always felt, the resulting inner branching / nesting complexity was not worth the flatter code...

If you feel very strongly about this, I can try and dig deeper, but would prefer to just leave it as is.